components/supervisor/pkg/ports/exposed-ports.go

```diff
@@ -105,22 +132,41 @@ func (g *GitpodExposedPorts) Observe(ctx context.Context) (<-chan []ExposedPort,
 			errchan <- err
 			return
 		}
+		mixin := func(localExposedPort []uint32, serverExposePort []*gitpod.WorkspaceInstancePort) []ExposedPort {
+			res := make(map[uint32]ExposedPort)
+			for _, port := range g.localExposedPort {
+				res[port] = ExposedPort{
+					LocalPort: port,
+					Public:    false,
+					URL:       g.getPortUrl(port),
+				}
+			}
+
+			for _, p := range serverExposePort {
+				res[uint32(p.Port)] = ExposedPort{
+					LocalPort: uint32(p.Port),
+					Public:    p.Visibility == "public",
+					URL:       p.URL,
+				}
+			}
+			exposedPort := make([]ExposedPort, 0, len(res))
+			for _, p := range res {
+				exposedPort = append(exposedPort, p)
+			}
+			return exposedPort
```

Comment:
Why not push to array directly? is it because we are not filtering public ports in `localExposedPort`?